src/NBomber/Domain/Domain.fs

```diff
@@ -62,221 +50,228 @@ type StepListenerChannel() =
                         timer.Stop()
                         let latency = Convert.ToInt64(timer.Elapsed.TotalMilliseconds)
-                        return (resp, latency)        
-        
+                        return (resp, latency)
+
         | Listener l -> let listener = l.Listeners.Get(req.CorrelationId)
                         let! resp = listener.GetResponse()
                         timer.Stop()
                         let latency = Convert.ToInt64(timer.Elapsed.TotalMilliseconds)
                         return (resp, latency)
-        
+
         | Pause time -> do! Task.Delay(time)
                         return (Response.Ok(req), int64(0))
     }
 
     let runSteps (steps: Step[], correlationId: CorrelationId,
                   latencies: List<List<Response*Latency>>,
                   exceptions: List<Option<exn>*ExceptionCount>, ct: CancellationToken) = task {
-        
-        do! Task.Delay(10)        
+
+        do! Task.Delay(10)
         let timer = Stopwatch()
 
         while not ct.IsCancellationRequested do
-            
+
             let mutable request = { CorrelationId = correlationId; Payload = null }
             let mutable skipStep = false
             let mutable stepIndex = 0
 
             for st in steps do
                 if not skipStep then
                     try
-                        let! (response,latency) = Step.execStep(st, request, timer)
-                            
-                        if not(Step.isPause st) then                        
+                        let! (response,latency) = execStep(st, request, timer)
+
+                        if not(isPause st) then
                             latencies.[stepIndex].Add(response,latency)
 
-                        if response.IsOk then 
+                        if response.IsOk then
                             request <- { request with Payload = response.Payload }
                             stepIndex <- stepIndex + 1
                         else
                             skipStep <- true
-                        
+
                     with ex -> let (_, exCount) = exceptions.[stepIndex]
                                let newCount = exCount + 1
                                exceptions.[stepIndex] <- (Some(ex), newCount)
                                skipStep <- true
     }
 
+type TestFlow = {
+    FlowName: FlowName
+    Steps: Step[]
+    CorrelationIds: Set<CorrelationId>
+}
+
+type Scenario = {
+    ScenarioName: string
+    InitStep: RequestStep option
+    TestFlows: TestFlow[]
+    Duration: TimeSpan
+    Assertions: Assertion[]
+}
+
+module internal Constants =
+
+    [<Literal>]
+    let WarmUpId = "warm_up_flow"
+
+    [<Literal>]
+    let InitId = "init_step"
+
 module internal TestFlow =
-    
-    let create (flowIndex, config: Contracts.TestFlow) =
+    let create (flowIndex, config: NBomber.Contracts.TestFlow) =
+        let initFlow (flow: TestFlow) =
```

Comment:
is it create function will produce side effects now? what the benefit?

Reply:
The `create` function does exactly what it was doing before. I've just moved the functions it was dependent upon in an attempt to keep related code close together. 

Also, I don't see that anyone need to access the `initFlow`, `initListeners` and `createCorrelationId` from outside the module so, I've encapsulated them using [nested functions](https://fsharpforfunandprofit.com/posts/organizing-functions/).

Reply:
what about testing?

Reply:
Sure, when we write unit tests we can take the nested functions out. In general, I think the point of nested functions is that they're simple and could be tested by testing the outer function itself.

The whole idea here was to not let a function pollute the module but I think we need to find a right balance on case-by-case basis. So, if you feel some of the above functions should not be nested, let me know and I can make the corresponding changes.